### PR TITLE
TNS PoC: TEM-68

### PR DESCRIPTION
**Pull request by postables**
## :construction_worker: Purpose
Proof of concept for Temporal Name Server (TNS)
Details about TNS can be seen [here](https://rtradetechnologies.atlassian.net/wiki/spaces/TEM/pages/16875550/Temporal+Name+Server+-+Technical+Specification+v1) note that this is not all of what is implemented in this PR.

## :rocket: Changes
Two new database models (record, zone)
Basic libp2p "client" and "server" for TNS
Additional API calls that can be used to lookup Records, and Zones from database
Additional API call to perform a `DagGet` call against IPFS
Updated utils test

## :warning: Breaking Changes
None

### CODE REVIEW

File: api/api.go [hunk 1]
```diff
@@ -250,6 +277,7 @@ func (api *API) setupRoutes() {
 		ipfs.GET("/pins", api.getLocalPins)                        // admin locked
 		ipfs.GET("/check-for-pin/:hash", api.checkLocalNodeForPin) // admin locked
 		ipfs.GET("/object-stat/:key", api.getObjectStatForIpfs)
+		ipfs.GET("/dag/get/:hash", api.getDagObject)
```
[the remark this answers is not in the document]
Comment (postables) [reply api/api.go]: It definitely doesn't I will change.

File: api/routes_rtfs.go [hunk 2]
```diff
@@ -372,6 +372,17 @@ func (api *API) checkLocalNodeForPin(c *gin.Context) {
 	Respond(c, http.StatusOK, gin.H{"response": present})
 }
 
+// GetDagObject is used to retrieve an IPLD object from ipfs
+func (api *API) getDagObject(c *gin.Context) {
+	hash := c.Param("hash")
+	var out interface{}
+	if err := api.ipfs.Shell.DagGet(hash, &out); err != nil {
+		api.LogError(err, eh.IPFSDagGetError)(c, http.StatusBadRequest)
```
[the remark this answers is not in the document]
Comment (postables) [reply api/routes_rtfs.go]: IIRC the only thing it should be is a bad request. If you request a non-existent IPLD object, the request will time-out

File: api/routes_tns.go [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+		}
+	}
+	req := queue.RecordCreation{
+		ZoneName:      zoneName,
+		RecordName:    recordName,
+		RecordKeyName: recordKeyName,
+		UserName:      username,
+	}
+	if len(intf) > 0 {
+		req.MetaData = intf
```
[the remark this answers is not in the document]
Comment (postables) [reply api/routes_tns.go]: Right!

File: api/routes_tns.go [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+	if !exists {
+		FailWithMissingField(c, "zone_name")
+		return
+	}
+	zoneManagerKeyName, exists := c.GetPostForm("zone_manager_key_name")
+	if !exists {
+		FailWithMissingField(c, "zone_manager_key_name")
+		return
+	}
+	zoneKeyName, exists := c.GetPostForm("zone_key_name")
```
[the remark this answers is not in the document]
Comment (postables) [reply api/routes_tns.go]: We definitely do, I'll add this to JIRA

File: tns/client.go [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+	// retrieve the dag from ipfs
+	var intf interface{}
+	if err = rtfsManager.Shell.DagGet(latestZoneHash, &intf); err != nil {
+		return nil, err
+	}
+	return intf, nil
+}
+
+// RecordRequest is a call used to request a record from TNS
+func (c *Client) RecordRequest(peerID peer.ID, req *RecordRequest) error {
```
[the remark this answers is not in the document]
Comment (postables) [reply tns/client.go]: Done!

File: models/tns.go [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+func (zm *ZoneManager) CheckIfRecordExistsInZone(zoneName, recordName, username string) (bool, error) {
+	z, err := zm.FindZoneByNameAndUser(zoneName, username)
+	if err != nil {
+		return false, err
+	}
+	if len(z.RecordNames) == 0 {
+		return false, nil
+	}
+	for _, v := range z.RecordNames {
+		if v == recordName {
```
[the remark this answers is not in the document]
Comment (postables) [reply models/tns.go]: Realistically it shouldn't be many, I would imagine at most 5-10. If it stars getting extremely large we may want to look into am apping

File: tns/client.go [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+	defaultZoneKeyName        = "postables-testkeydemo"
+	defaultZoneUserName       = "postables"
+	defaultRecordName         = "myrecord"
+	defaultRecordKeyName      = "postables-testkeydemo2"
+	defaultRecordUserName     = "postables"
+	dev                       = true
+)
+
+// GenerateTNSClient is used to generate a TNS Client
+func GenerateTNSClient(genPK bool, pk ...ci.PrivKey) (*Client, error) {
```
[the remark this answers is not in the document]
Comment (postables) [reply tns/client.go]: Good point, I think this was from early on when I was writing the code and forgot to clean it up. Tbh I don't have any other explanation haha

File: cmd/temporal/main.go [hunk 8]
```diff
@@ -24,6 +28,95 @@ var (
 )
 
 var commands = map[string]cmd.Cmd{
+	"tns": cmd.Cmd{
+		Blurb:         "run a tns daemon or client",
```
[the remark this answers is not in the document]
Comment (bobheadxi) [reply cmd/temporal/main.go]: sounds good!

File: models/record.go [hunk 9]
```diff
@@ -0,0 +1,93 @@
+package models
```
Comment (bobheadxi) [inline models/record.go]: @postables once this PR lands should I have another crack at extracting the database stuff into RTradeLtd/database? a bit worried about the deviation between the packages in Temporal growing too large to perform a non-trivial merge upstream

Comment (postables) [reply models/record.go]: @bobheadxi  yes that sounds like a great idea! I imagine now that we got rid of that stinking utility that was causing issues it should be easy to do.

**Pull request merged**